File: src/native/schemas/shellVariable.ts
import { z } from "zod";

/**
 * Shell start-up variable wire types (ADR-0042).
 *
 * The value travels because the edit dialog has to prefill it — asking someone
 * to retype an address they cannot see is not an edit, it is a guess — and
 * because the write compares it to refuse a line that changed in the meantime.
 * The path travels already shortened to `~/…`, and is display only: a write
 * names the variable and the line, and the backend re-locates the file itself.
 */

export const shellVariableLocationSchema = z
  .object({
    variable: z.string(),
    path: z.string(),
    line: z.number().int().positive(),
    value: z.string(),
    /** False for a command substitution or interpolation, where replacing the
     *  text with a literal would change what the line does. */
    editable: z.boolean(),
  })
  .strict();

export const shellVariableUpdateSchema = z
  .object({
    variable: z.string().min(1),
    line: z.number().int().positive(),
    expectedValue: z.string(),
    /** A line break would split one assignment into two statements. */
    newValue: z
      .string()
      .max(4096)
      .regex(/^[^\n\r]*$/),
  })
  .strict();

export const shellVariableWrittenSchema = z
  .object({
    path: z.string(),
    line: z.number().int().positive(),
  })
  .strict();

export const shellVariableLocationListSchema = z.array(
  shellVariableLocationSchema,
);

export type ShellVariableLocation = z.infer<typeof shellVariableLocationSchema>;
export type ShellVariableUpdate = z.infer<typeof shellVariableUpdateSchema>;
export type ShellVariableWritten = z.infer<typeof shellVariableWrittenSchema>;

File: src-tauri/src/domain/shell_variable.rs
//! Editing a connection variable a shell profile sets (ADR-0042).

use serde::{Deserialize, Serialize};

/// Where a connection variable is written down, and whether this product can
/// safely change it.
#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShellVariableLocation {
    pub variable: String,
    /// The start-up file, shortened to `~/…` for display. Never an absolute
    /// path outside the home directory in the common case, and never used by
    /// the renderer to address a write: the write re-locates the line itself.
    pub path: String,
    /// 1-based, so it reads the same as it does in an editor.
    pub line: u32,
    /// The value on that line, so the dialog can prefill it and so the write
    /// can refuse if the line has changed since.
    pub value: String,
    /// False when the line is a command substitution or an interpolation,
    /// where replacing the text with a literal would change what the line
    /// does rather than what it says. The edit action is withheld.
    pub editable: bool,
}

/// Redacted: the value is a credential as often as it is an address.
impl std::fmt::Debug for ShellVariableLocation {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ShellVariableLocation")
            .field("variable", &self.variable)
            .field("path", &"<path>")
            .field("line", &self.line)
            .field("value", &"<redacted>")
            .field("editable", &self.editable)
            .finish()
    }
}

/// One replacement, carrying what the caller last saw so a line that moved on
/// cannot be overwritten blindly.
#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShellVariableUpdate {
    pub variable: String,
    pub line: u32,
    pub expected_value: String,
    pub new_value: String,
}

impl std::fmt::Debug for ShellVariableUpdate {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ShellVariableUpdate")
            .field("variable", &self.variable)
            .field("line", &self.line)
            .field("expected_value", &"<redacted>")
            .field("new_value", &"<redacted>")
            .finish()
    }
}

/// What changed, for the dialog to report without guessing.
#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShellVariableWritten {
    /// The file that was rewritten, shortened for display.
    pub path: String,
    pub line: u32,
}

impl std::fmt::Debug for ShellVariableWritten {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ShellVariableWritten")
            .field("path", &"<path>")
            .field("line", &self.line)
            .finish()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_location_never_renders_its_value() {
        let location = ShellVariableLocation {
            variable: "ANTHROPIC_AUTH_TOKEN".to_string(),
            path: "~/.config/zsh/secrets.zsh".to_string(),
            line: 14,
            value: "sk-must-not-be-logged".to_string(),
            editable: true,
        };
        let rendered = format!("{location:?}");
        assert!(!rendered.contains("sk-must-not-be-logged"), "{rendered}");
        assert!(!rendered.contains("secrets.zsh"), "{rendered}");
        // The variable name and the line number are safe and are what a log
        // needs in order to be useful at all.
        assert!(rendered.contains("ANTHROPIC_AUTH_TOKEN"), "{rendered}");
        assert!(rendered.contains("14"), "{rendered}");
    }

    #[test]
    fn an_update_never_renders_either_value() {
        let update = ShellVariableUpdate {
            variable: "ANTHROPIC_BASE_URL".to_string(),
            line: 3,
            expected_value: "https://old.example.test".to_string(),
            new_value: "https://new.example.test".to_string(),
        };
        let rendered = format!("{update:?}");
        assert!(!rendered.contains("example.test"), "{rendered}");
    }
}
